File: gui/menu/workspace/workspace_copy_into.cpp
#include "workspace_copy_into.h"


MenuWorkspaceCopyInto::MenuWorkspaceCopyInto(QWidget* parent) : 
    IodeSettingsDialog(parent)
{
    setupUi(this);

    QList<QString> qIodeTypes;
    for(const std::string& iode_type: vIodeTypes) qIodeTypes << QString::fromStdString(iode_type);

    wComboBoxObjectType = new WrapperComboBox(label_object_type->text(), *comboBox_object_type, REQUIRED_FIELD, qIodeTypes);
    wFileChooserInputFile = new WrapperFileChooser(label_input_file->text(), *fileChooser_input_file, REQUIRED_FIELD, I_COMMENTS_FILE, EXISTING_FILE);
    wSampleEditSampleFrom = new WrapperSampleEdit(label_sample_from->text(), *sampleEdit_sample_from, OPTIONAL_FIELD);
    wSampleEditSampleTo = new WrapperSampleEdit(sampleEdit_sample_to->text(), *sampleEdit_sample_to, OPTIONAL_FIELD);
    wTextEditObjectNames = new WrapperQTextEdit(label_object_names->text(), *textEdit_object_names, OPTIONAL_FIELD);

    mapFields["ObjectType"] = wComboBoxObjectType;
    mapFields["InputFile"] = wFileChooserInputFile;
    mapFields["SampleFrom"] = wSampleEditSampleFrom;
    mapFields["SampleTo"] = wSampleEditSampleTo;
    mapFields["ObjectNames"] = wTextEditObjectNames;

    connect(&wComboBoxObjectType->getQField(), &QComboBox::currentIndexChanged, this, &MenuWorkspaceCopyInto::updateFileChooser);

    className = "MENU_WORKSPACE_COPY_INTO";
    loadSettings();
}

MenuWorkspaceCopyInto::~MenuWorkspaceCopyInto()
{
    delete wComboBoxObjectType;
    delete wFileChooserInputFile;
    delete wSampleEditSampleFrom;
    delete wSampleEditSampleTo;
    delete wTextEditObjectNames;
}

void MenuWorkspaceCopyInto::copy_into_workspace()
{
    try
    {
        int iodeType = wComboBoxObjectType->extractAndVerify();
        std::string input_file = wFileChooserInputFile->extractAndVerify().toStdString();
        std::string from = wSampleEditSampleFrom->extractAndVerify().toStdString();
        std::string to = wSampleEditSampleTo->extractAndVerify().toStdString();
        std::string object_names = wTextEditObjectNames->extractAndVerify().toStdString();

        switch (iodeType)
        {
        case I_COMMENTS:
            Comments.copy_into(input_file, object_names);
            break;
        case I_EQUATIONS:
            Equations.copy_into(input_file, object_names);
            break;
        case I_IDENTITIES:
            Identities.copy_into(input_file, object_names);
            break;
        case I_LISTS:
            Lists.copy_into(input_file, object_names);
            break;
        case I_SCALARS:
            Scalars.copy_into(input_file, object_names);
            break;
        case I_TABLES:
            Tables.copy_into(input_file, object_names);
            break;
        case I_VARIABLES:
            Variables.copy_into(input_file, from, to, object_names);
            break;
        default:
            break;
        }
    }
    catch (const std::exception& e)
    {
        QMessageBox::critical(this, tr("ERROR"), tr(e.what()));
        return;
    }

    this->accept();
}

void MenuWorkspaceCopyInto::help()
{
	QDesktopServices::openUrl(url_manual);
}

File: gui/menu/workspace/workspace_copy_into.py
from PySide6.QtCore import Slot
from PySide6.QtWidgets import QWidget, QMessageBox

from settings import MixinSettingsDialog
from .ui_workspace_copy_into import Ui_MenuWorkspaceCopyInto


from iode import (IodeType, IodeFileType, comments, equations, identities, 
                  lists, scalars, tables, variables)


class MenuWorkspaceCopyInto(MixinSettingsDialog):
    def __init__(self, parent: QWidget=None):
        super().__init__(parent)
        self.ui = Ui_MenuWorkspaceCopyInto()
        self.ui.setupUi(self)
        self.prepare_settings(self.ui)

        self.v_files_types = list(IodeFileType)
        self.v_iode_types = list(IodeType)
        v_iode_type_names = [iode_type.name for iode_type in self.v_iode_types]
        self.ui.comboBox_object_type.addItems(v_iode_type_names)    
        self.ui.comboBox_object_type.currentIndexChanged.connect(self.update_file_chooser)

        self.load_settings()

    @Slot()
    def copy_into_workspace(self):
        try:
            iode_type: IodeType = self.v_iode_types[self.ui.comboBox_object_type.currentIndex()]
            input_file: str = self.ui.fileChooser_input_file.filepath
            from_period: str = self.ui.sampleEdit_sample_from.text()
            to_period: str = self.ui.sampleEdit_sample_to.text()
            object_names: str = ";".join(self.ui.textEdit_object_names.toPlainText().splitlines()).strip()

            if not len(input_file):
                raise ValueError("'input file' is empty")
            
            if not len(from_period):
                from_period = None
            if not len(to_period):
                to_period = None

            if iode_type == IodeType.COMMENTS:
                comments.copy_from(input_file, object_names)
            elif iode_type == IodeType.EQUATIONS:
                equations.copy_from(input_file, object_names)
            elif iode_type == IodeType.IDENTITIES:
                identities.copy_from(input_file, object_names)
            elif iode_type == IodeType.LISTS:
                lists.copy_from(input_file, object_names)
            elif iode_type == IodeType.SCALARS:
                scalars.copy_from(input_file, object_names)
            elif iode_type == IodeType.TABLES:
                tables.copy_from(input_file, object_names)
            elif iode_type == IodeType.VARIABLES:
                variables.copy_from(input_file, from_period, to_period, object_names)

            self.accept()
        except Exception as e:
            QMessageBox.warning(self, "WARNING", "Failed to copy into workspace.\n" + str(e))

    @Slot(int)
    def update_file_chooser(self, i_iode_type: int):
        self.ui.fileChooser_input_file.enum_file_type = self.v_files_types[i_iode_type]
